jax/_src/dispatch.py

```diff
@@ -430,12 +416,10 @@ def jaxpr_has_pmap(jaxpr):
   return False
 
 def jaxpr_has_bints(jaxpr: core.Jaxpr) -> bool:
-  return (any(type(d) is core.Var for v in jaxpr.invars
-              if type(v.aval) is core.DShapedArray for d in v.aval.shape) or
-          any(type(d) is core.Var
+  return (any(type(v.aval) is core.AbstractBInt for v in jaxpr.invars) or
+          any(type(v.aval) is core.AbstractBInt
               for j in itertools.chain([jaxpr], core.subjaxprs(jaxpr))
```

Comment:
Could be `[jaxpr, *core.subjaxprs(jaxpr)]`

Reply:
I wrote it this way b/c in theory `itertools.chain` can be more memory efficient, but OTOH that seems like not very important here. WDYT?

Reply:
SGTM